File: src/commands/repo.ts
import { Command } from "commander";
import { requireAuth } from "../api.js";
import { success, error, output, info } from "../utils/output.js";

function parseRepoRef(ref: string): { store: string; repo: string } {
  const parts = ref.split("/");
  if (parts.length !== 2) {
    error(`Invalid format: "${ref}"`);
    info("Expected format: <store>/<repo>");
    info("Examples:");
    info("  scraps repo show alice/my-project");
    info("  scraps repo delete mystore/backend");
    process.exit(1);
  }
  return { store: parts[0], repo: parts[1] };
}

export function registerRepoCommands(program: Command): void {
  const repo = program
    .command("repo")
    .description("Manage repositories")
    .addHelpText("after", `
Examples:
  scraps repo list                    # List repos in all your stores
  scraps repo list alice              # List repos in store "alice"
  scraps repo create alice/my-app     # Create repo "my-app" in store "alice"
  scraps repo show alice/my-app       # Show repo details
  scraps repo delete alice/my-app     # Delete a repo
`);

  repo
    .command("create <store/repo>")
    .description("Create a new repository")
    .addHelpText("after", `
Examples:
  scraps repo create alice/my-project
  scraps repo create myteam/backend-api
`)
    .action(async (ref) => {
      const client = requireAuth();
      const { store, repo: name } = parseRepoRef(ref);
      try {
        const result = await client.post(`/api/v1/stores/${store}/repos`, {
          name,
        });
        success(`Repository '${store}/${name}' created`);
        output(result.repo || result);
      } catch (e: any) {
        error(`Failed to create repo: ${e.message}`);
        process.exit(1);
      }
    });

  repo
    .command("list [store]")
    .description("List repositories (in a store, or all stores if omitted)")
    .addHelpText("after", `
Examples:
  scraps repo list              # List all repos across all your stores
  scraps repo list alice        # List repos only in store "alice"
`)
    .action(async (store) => {
      const client = requireAuth();
      try {
        if (store) {
          // List repos in specific store
          const result = await client.get(`/api/v1/stores/${store}/repos`);
          const repos = result.repos || result;
          if (repos.length === 0) {
            info(`No repositories in store "${store}"`);
            return;
          }
          output(repos, {
            headers: ["Name", "Default Branch", "Created"],
            rows: repos.map((r: any) => [
              r.name,
              r.default_branch || "main",
              new Date(r.created_at).toLocaleDateString(),
            ]),
          });
        } else {
          // List repos across all stores
          const storesResult = await client.get("/api/v1/stores");
          const stores = storesResult.stores || storesResult;

          let allRepos: any[] = [];
          for (const s of stores) {
            try {
              const result = await client.get(`/api/v1/stores/${s.slug}/repos`);
              const repos = result.repos || result;
              for (const r of repos) {
                allRepos.push({ ...r, store: s.slug });
              }
            } catch {
              // Skip stores we can't access
            }
          }

          if (allRepos.length === 0) {
            info("No repositories found");
            info("Create one with: scraps repo create <store>/<name>");
            return;
          }

          output(allRepos, {
            headers: ["Store", "Name", "Default Branch"],
            rows: allRepos.map((r: any) => [
              r.store,
              r.name,
              r.default_branch || "main",
            ]),
          });
        }
      } catch (e: any) {
        error(`Failed to list repos: ${e.message}`);
        process.exit(1);
      }
    });

  repo
    .command("show <store/repo>")
    .description("Show repository details")
    .addHelpText("after", `
Examples:
  scraps repo show alice/my-project
`)
    .action(async (ref) => {
      const client = requireAuth();
      const { store, repo: name } = parseRepoRef(ref);
      try {
        const result = await client.get(`/api/v1/stores/${store}/repos/${name}`);
        const r = result.repo || result;
        output(r, {
          headers: ["Field", "Value"],
          rows: [
            ["ID", r.id],
            ["Name", r.name],
            ["Store", store],
            ["Default Branch", r.default_branch || "main"],
            ["Created", new Date(r.created_at).toLocaleDateString()],
          ],
        });
      } catch (e: any) {
        error(`Failed to get repo: ${e.message}`);
        process.exit(1);
      }
    });

  repo
    .command("delete <store/repo>")
    .description("Delete a repository")
    .addHelpText("after", `
Examples:
  scraps repo delete alice/old-project
`)
    .action(async (ref) => {
      const client = requireAuth();
      const { store, repo: name } = parseRepoRef(ref);
      try {
        await client.delete(`/api/v1/stores/${store}/repos/${name}`);
        success(`Repository '${store}/${name}' deleted`);
      } catch (e: any) {
        error(`Failed to delete repo: ${e.message}`);
        process.exit(1);
      }
    });

  // Collaborators subcommand
  const collaborators = repo
    .command("collaborators")
    .description("Manage repository collaborators")
    .addHelpText("after", `
Examples:
  scraps repo collaborators list alice/my-project
  scraps repo collaborators add alice/my-project bob --role write
  scraps repo collaborators remove alice/my-project bob
`);

  collaborators
    .command("list <store/repo>")
    .description("List repository collaborators")
    .action(async (ref) => {
      const client = requireAuth();
      const { store, repo: name } = parseRepoRef(ref);
      try {
        const result = await client.get(
          `/api/v1/stores/${store}/repos/${name}/collaborators`
        );
        const collabs = result.collaborators || result;
        if (collabs.length === 0) {
          info("No collaborators");
          return;
        }
        output(collabs, {
          headers: ["Username", "Role", "Added"],
          rows: collabs.map((c: any) => [
            c.username,
            c.role,
            new Date(c.created_at).toLocaleDateString(),
          ]),
        });
      } catch (e: any) {
        error(`Failed to list collaborators: ${e.message}`);
        process.exit(1);
      }
    });

  collaborators
    .command("add <store/repo> <username>")
    .description("Add a collaborator to the repository")
    .option("-r, --role <role>", "Role: read, write, or admin", "read")
    .action(async (ref, username, opts) => {
      const client = requireAuth();
      const { store, repo: name } = parseRepoRef(ref);
      try {
        await client.post(`/api/v1/stores/${store}/repos/${name}/collaborators`, {
          username,
          role: opts.role,
        });
        success(`Added ${username} to ${store}/${name} with ${opts.role} access`);
      } catch (e: any) {
        error(`Failed to add collaborator: ${e.message}`);
        process.exit(1);
      }
    });

  collaborators
    .command("remove <store/repo> <username>")
    .description("Remove a collaborator from the repository")
    .action(async (ref, username) => {
      const client = requireAuth();
      const { store, repo: name } = parseRepoRef(ref);
      try {
        const result = await client.get(
          `/api/v1/stores/${store}/repos/${name}/collaborators`
        );
        const collabs = result.collaborators || result;
        const collab = collabs.find((c: any) => c.username === username);
        if (!collab) {
          error(`Collaborator '${username}' not found`);
          process.exit(1);
        }
        await client.delete(
          `/api/v1/stores/${store}/repos/${name}/collaborators/${collab.id}`
        );
        success(`Removed ${username} from ${store}/${name}`);
      } catch (e: any) {
        error(`Failed to remove collaborator: ${e.message}`);
        process.exit(1);
      }
    });
}

File: internal/cli/repo.go
package cli

import (
	"fmt"

	"github.com/spf13/cobra"

	"github.com/morrisclay/scraps-cli/internal/api"
	"github.com/morrisclay/scraps-cli/internal/config"
	"github.com/morrisclay/scraps-cli/internal/tui/components"
)

func newRepoCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "repo",
		Short: "Manage repositories",
	}

	cmd.AddCommand(newRepoListCmd())
	cmd.AddCommand(newRepoCreateCmd())
	cmd.AddCommand(newRepoShowCmd())
	cmd.AddCommand(newRepoDeleteCmd())
	cmd.AddCommand(newRepoCollaboratorsCmd())

	return cmd
}

func newRepoListCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list [store]",
		Short: "List repositories",
		Long:  "List repositories. If store is specified, lists repos in that store. Otherwise lists all accessible repos.",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			var repos []struct {
				Store string
				Name  string
				ID    string
				CreatedAt string
			}

			if len(args) > 0 {
				// List repos in specific store
				storeRepos, err := client.ListRepos(args[0])
				if err != nil {
					return err
				}
				for _, r := range storeRepos {
					repos = append(repos, struct {
						Store     string
						Name      string
						ID        string
						CreatedAt string
					}{args[0], r.Name, r.ID, r.CreatedAt})
				}
			} else {
				// List all repos
				stores, err := client.ListStores()
				if err != nil {
					return err
				}
				for _, store := range stores {
					storeRepos, err := client.ListRepos(store.Slug)
					if err != nil {
						continue
					}
					for _, r := range storeRepos {
						repos = append(repos, struct {
							Store     string
							Name      string
							ID        string
							CreatedAt string
						}{store.Slug, r.Name, r.ID, r.CreatedAt})
					}
				}
			}

			if len(repos) == 0 {
				info("No repositories found")
				return nil
			}

			if config.GetOutputFormat() == "json" {
				outputJSON(repos)
			} else {
				// Interactive mode - use searchable list
				if isInteractive() && len(args) == 0 {
					items := make([]components.SearchListItem, len(repos))
					for i, r := range repos {
						items[i] = components.NewSearchListItem(
							formatStoreRepo(r.Store, r.Name),
							fmt.Sprintf("Created: %s", formatDate(r.CreatedAt)),
							r,
						)
					}

					selected, err := components.RunSearchList("Select Repository", items)
					if err != nil {
						return err
					}
					if selected != nil {
						fmt.Printf("Selected: %s\n", selected.Title())
					}
					return nil
				}

				// Table output
				headers := []string{"REPOSITORY", "CREATED"}
				rows := make([][]string, len(repos))
				for i, r := range repos {
					rows[i] = []string{formatStoreRepo(r.Store, r.Name), formatDate(r.CreatedAt)}
				}
				outputTable(headers, rows)
			}
			return nil
		},
	}
	return cmd
}

func newRepoCreateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "create <store/repo>",
		Short: "Create a new repository",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, name, err := parseStoreRepo(args[0])
			if err != nil {
				return err
			}

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			repo, err := client.CreateRepo(store, name)
			if err != nil {
				return err
			}

			if config.GetOutputFormat() == "json" {
				outputJSON(repo)
			} else {
				success(fmt.Sprintf("Repository '%s/%s' created", store, repo.Name))
			}
			return nil
		},
	}
	return cmd
}

func newRepoShowCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "show <store/repo>",
		Short: "Show repository details",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, name, err := parseStoreRepo(args[0])
			if err != nil {
				return err
			}

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			repo, err := client.GetRepo(store, name)
			if err != nil {
				return err
			}

			if config.GetOutputFormat() == "json" {
				outputJSON(repo)
			} else {
				fmt.Printf("Name:           %s\n", repo.Name)
				fmt.Printf("Store:          %s\n", store)
				fmt.Printf("ID:             %s\n", repo.ID)
				fmt.Printf("Default Branch: %s\n", repo.DefaultBranch)
				fmt.Printf("Created:        %s\n", formatDateTime(repo.CreatedAt))
			}
			return nil
		},
	}
	return cmd
}

func newRepoDeleteCmd() *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:   "delete <store/repo>",
		Short: "Delete a repository",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, name, err := parseStoreRepo(args[0])
			if err != nil {
				return err
			}

			// Confirm deletion
			if !force && isInteractive() {
				confirmed, err := components.RunConfirm(
					"Delete Repository",
					fmt.Sprintf("Are you sure you want to delete '%s/%s'?\nThis cannot be undone.", store, name),
					true,
				)
				if err != nil {
					return err
				}
				if !confirmed {
					info("Deletion cancelled")
					return nil
				}
			}

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			if err := client.DeleteRepo(store, name); err != nil {
				return err
			}

			success(fmt.Sprintf("Repository '%s/%s' deleted", store, name))
			return nil
		},
	}

	cmd.Flags().BoolVarP(&force, "force", "f", false, "Skip confirmation prompt")
	return cmd
}

// --- Repository Collaborators ---

func newRepoCollaboratorsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "collaborators",
		Short: "Manage repository collaborators",
	}

	cmd.AddCommand(newRepoCollaboratorsListCmd())
	cmd.AddCommand(newRepoCollaboratorsAddCmd())
	cmd.AddCommand(newRepoCollaboratorsRemoveCmd())

	return cmd
}

func newRepoCollaboratorsListCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list <store/repo>",
		Short: "List collaborators of a repository",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, name, err := parseStoreRepo(args[0])
			if err != nil {
				return err
			}

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			collabs, err := client.ListCollaborators(store, name)
			if err != nil {
				return err
			}

			if len(collabs) == 0 {
				info("No collaborators found")
				return nil
			}

			if config.GetOutputFormat() == "json" {
				outputJSON(collabs)
			} else {
				headers := []string{"USERNAME", "ROLE", "ADDED"}
				rows := make([][]string, len(collabs))
				for i, c := range collabs {
					rows[i] = []string{c.Username, c.Role, formatDate(c.CreatedAt)}
				}
				outputTable(headers, rows)
			}
			return nil
		},
	}
	return cmd
}

func newRepoCollaboratorsAddCmd() *cobra.Command {
	var role string

	cmd := &cobra.Command{
		Use:   "add <store/repo> <username>",
		Short: "Add a collaborator to a repository",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, name, err := parseStoreRepo(args[0])
			if err != nil {
				return err
			}
			username := args[1]

			if role == "" {
				role = "read"
			}

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			collab, err := client.AddCollaborator(store, name, username, role)
			if err != nil {
				return err
			}

			if config.GetOutputFormat() == "json" {
				outputJSON(collab)
			} else {
				success(fmt.Sprintf("Added %s to %s/%s with role %s", username, store, name, collab.Role))
			}
			return nil
		},
	}

	cmd.Flags().StringVarP(&role, "role", "r", "read", "Collaborator role (read, write, admin)")
	return cmd
}

func newRepoCollaboratorsRemoveCmd() *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:   "remove <store/repo> <username>",
		Short: "Remove a collaborator from a repository",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			store, name, err := parseStoreRepo(args[0])
			if err != nil {
				return err
			}
			username := args[1]

			// Confirm removal
			if !force && isInteractive() {
				confirmed, err := components.RunConfirm(
					"Remove Collaborator",
					fmt.Sprintf("Remove '%s' from '%s/%s'?", username, store, name),
					false,
				)
				if err != nil {
					return err
				}
				if !confirmed {
					info("Removal cancelled")
					return nil
				}
			}

			client, err := api.NewClientFromConfig("")
			if err != nil {
				return err
			}

			// Find collaborator ID
			collabs, err := client.ListCollaborators(store, name)
			if err != nil {
				return err
			}

			var collabID string
			for _, c := range collabs {
				if c.Username == username {
					collabID = c.ID
					break
				}
			}

			if collabID == "" {
				return fmt.Errorf("collaborator '%s' not found in '%s/%s'", username, store, name)
			}

			if err := client.RemoveCollaborator(store, name, collabID); err != nil {
				return err
			}

			success(fmt.Sprintf("Removed %s from %s/%s", username, store, name))
			return nil
		},
	}

	cmd.Flags().BoolVarP(&force, "force", "f", false, "Skip confirmation prompt")
	return cmd
}
